{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:55:20.364449Z",
     "end_time": "2023-05-11T15:55:20.471005Z"
    },
    "id": "o3XLG2RjlCBE"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import OrderedDict\n",
    "from IPython.display import display\n",
    "import random\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import numba\n",
    "from numba import jit, prange, njit, types\n",
    "import sys\n",
    "from numba import int64, boolean, types, typed\n",
    "from numba.experimental import jitclass\n",
    "from numba.core.errors import NumbaDeprecationWarning, NumbaPendingDeprecationWarning\n",
    "from numba.core.errors import NumbaWarning\n",
    "import warnings\n",
    "from numba.typed import Dict, List\n",
    "import hashlib\n",
    "\n",
    "warnings.simplefilter('ignore', category=NumbaDeprecationWarning)\n",
    "warnings.simplefilter('ignore', category=NumbaPendingDeprecationWarning)\n",
    "warnings.simplefilter('ignore', category=NumbaWarning)\n",
    "\n",
    "#sys.setrecursionlimit(10**4)\n",
    "# import display\n",
    "def load_and_print_csvs_from_folders():\n",
    "    cwd = os.getcwd()\n",
    "    folders = [d for d in os.listdir(cwd) if os.path.isdir(os.path.join(cwd, d))]\n",
    "    data = {}\n",
    "    for folder in folders:\n",
    "        folder_path = os.path.join(cwd, folder)\n",
    "        for file in os.listdir(folder_path):\n",
    "            if file.endswith('.csv'):\n",
    "                file_path = os.path.join(folder_path, file)\n",
    "                df = pd.read_csv(file_path)\n",
    "                data[file] = df\n",
    "\n",
    "    return data\n",
    "\n",
    "def pretty_print_dict(dictionary, indent=0):\n",
    "    for key, value in dictionary.items():\n",
    "        if isinstance(value, dict):\n",
    "            print('  ' * indent + str(key) + ':')\n",
    "            pretty_print_dict(value, indent+1)\n",
    "        else:\n",
    "            print('  ' * indent + str(key) + ': ' + str(value))\n",
    "@njit\n",
    "def int_to_base36(value):\n",
    "    if value < 0:\n",
    "        value = -value\n",
    "    if value == 0:\n",
    "        return \"0\"\n",
    "\n",
    "    base36_digits = \"0123456789abcdefghijklmnopqrstuvwxyz\"\n",
    "    base36 = []\n",
    "    while value:\n",
    "        value, i = divmod(value, 36)\n",
    "        base36.append(base36_digits[i])\n",
    "\n",
    "    # Reversing the list\n",
    "    base36 = base36[::-1]\n",
    "\n",
    "    # Joining the list into a string\n",
    "    base36_str = \"\"\n",
    "    for char in base36:\n",
    "        base36_str += char\n",
    "\n",
    "    return base36_str\n",
    "\n",
    "@jit(nopython=False)\n",
    "def genotype_hash(genotype_param, precision=16):\n",
    "    # Sort genotype keys to ensure a consistent order\n",
    "    sorted_keys = sorted(genotype_param.keys())\n",
    "\n",
    "    # Create a string representation of the genotype\n",
    "    genotype_str = \"\"\n",
    "    for key in sorted_keys:\n",
    "        genotype_str += key + \":\"\n",
    "        #genotype_str += \",\".join(str(val) for val in genotype_param[key])\n",
    "        for i in range(len(genotype_param[key])):\n",
    "            genotype_str += \",\" + str(genotype_param[key][i])\n",
    "        genotype_str += \";\"\n",
    "\n",
    "    hash_object = hash(genotype_str)\n",
    "    str_hash = int_to_base36(hash_object)\n",
    "    length = min(precision, len(str_hash))\n",
    "\n",
    "    # get the first 16 characters of the hash\n",
    "    return str_hash[:length]\n",
    "\n",
    "@njit\n",
    "def deep_copy_genotype(genotype, empty_genotype):\n",
    "\n",
    "    for key in genotype:\n",
    "        values = genotype[key]\n",
    "        copied_values = values.copy()\n",
    "        empty_genotype[key] = copied_values\n",
    "\n",
    "    return empty_genotype\n",
    "#load_and_print_csvs_from_folders()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<start>: [[<expr>, ...], ...]\n",
      "<expr>: [[<number>, ...], [<var>, ...], [(, <op>, <expr1>, <expr1>, ), ...], [(, <uop>, <expr1>, ), ...], ...]\n",
      "<number>: [[<integer>, ...], ...]\n",
      "<integer>: [[(, <digit>, ), ...], [(, <non-zero-digit>, <digit>, ), ...], [(, <non-zero-digit>, <digit>, <digit>, ), ...], ...]\n",
      "<digit>: [[0, ...], [<non-zero-digit>, ...], ...]\n",
      "<non-zero-digit>: [[1, ...], [2, ...], [3, ...], [4, ...], [5, ...], [6, ...], [7, ...], [8, ...], [9, ...], ...]\n",
      "<op>: [['+', ...], ['-', ...], ['*', ...], ['/', ...], ...]\n",
      "<uop>: [['abs', ...], ...]\n",
      "<var>: [[x, ...], [y, ...], ...]\n",
      "<expr1>: [[<number>, ...], [<var>, ...], [(, <op>, <expr2>, <expr2>, ), ...], [(, <uop>, <expr2>, ), ...], ...]\n",
      "<expr2>: [[<number>, ...], [<var>, ...], [(, <op>, <expr3>, <expr3>, ), ...], [(, <uop>, <expr3>, ), ...], ...]\n",
      "<expr3>: [[<number>, ...], [<var>, ...], [(, <op>, <expr4>, <expr4>, ), ...], [(, <uop>, <expr4>, ), ...], ...]\n",
      "<expr4>: [[<number>, ...], [<var>, ...], ...]\n"
     ]
    }
   ],
   "source": [
    "def read_bnf_file(file_path):\n",
    "    # read file as string\n",
    "    with open(file_path, 'r') as f:\n",
    "        bnf_lines = f.read()\n",
    "    return bnf_lines\n",
    "\n",
    "def parse_bnf(grammar_str, depth=1, variables=1):\n",
    "    #productions = Dict.empty(key_type=types.unicode_type, value_type=types.ListType(types.ListType(types.unicode_type)))\n",
    "    productions = {}\n",
    "    grammar_str = re.sub(r'\\s*\\|\\s*', ' |', grammar_str)\n",
    "    lines = [line.strip() for line in grammar_str.split('\\n') if line.strip()]\n",
    "    PRODUCTION_SEPARATOR = '::='\n",
    "    RULE_SEPARATOR = '|'\n",
    "    RECURSIVE_PRODUCTIONS = ['<expr>']\n",
    "    for line in lines:\n",
    "        lhs, rhs = line.split(PRODUCTION_SEPARATOR)\n",
    "        lhs = lhs.strip()\n",
    "        rhs_productions = []\n",
    "        for rule in rhs.split(RULE_SEPARATOR):\n",
    "            rule_list = list(rule.strip().split())\n",
    "            rhs_productions.append(rule_list)\n",
    "\n",
    "        productions[lhs] = rhs_productions\n",
    "\n",
    "    def list_new_non_recursive_expressions(recursive_rule):\n",
    "        # create depth expressions\n",
    "        expressions = [recursive_rule]\n",
    "        for i in range(1, depth):\n",
    "            expr_rule = recursive_rule.replace('>', f'{i}>')\n",
    "            expressions.append(expr_rule)\n",
    "        return expressions\n",
    "\n",
    "    def filter_non_recursive_productions(recursive_rule):\n",
    "        non_recursive_productions = set()\n",
    "        for exp in productions[recursive_rule]:\n",
    "            if recursive_rule not in exp:\n",
    "                non_recursive_productions.add(tuple(exp))\n",
    "\n",
    "        return non_recursive_productions\n",
    "\n",
    "    def replace_expression_in_production(old_rule, old, new):\n",
    "        new_rule = list(old_rule)\n",
    "        for i, exp in enumerate(new_rule):\n",
    "            if new_rule[i] == old:\n",
    "                new_rule[i] = new\n",
    "        return new_rule\n",
    "\n",
    "    def fix_recursive_production(recursive_rule):\n",
    "        non_recursive_productions = filter_non_recursive_productions(recursive_rule)\n",
    "        if depth == 1:\n",
    "            productions[recursive_rule] = non_recursive_productions\n",
    "            return\n",
    "\n",
    "        new_expressions = list_new_non_recursive_expressions(recursive_rule)\n",
    "\n",
    "        all_production = set(tuple(p) for p in productions[recursive_rule])\n",
    "        recursive_productions = all_production - non_recursive_productions\n",
    "\n",
    "        for i in range(0, depth-1):\n",
    "            expr_rule = new_expressions[i]\n",
    "            new_productions = []\n",
    "            for rule in recursive_productions:\n",
    "                new_rule = replace_expression_in_production(rule, recursive_rule, new_expressions[i+1])\n",
    "                new_productions.append(new_rule)\n",
    "            productions[expr_rule] = [*non_recursive_productions, *new_productions]\n",
    "\n",
    "\n",
    "        productions[new_expressions[-1]] = list(non_recursive_productions)\n",
    "\n",
    "    for recursive_rule in RECURSIVE_PRODUCTIONS:\n",
    "        fix_recursive_production(recursive_rule)\n",
    "\n",
    "    variable_cases = ['x', 'y', 'z', 'w', 'k']\n",
    "    for var_index in range(variables+1):\n",
    "        terminal_vars = variable_cases[:var_index]\n",
    "        productions[\"<var>\"] = terminal_vars\n",
    "\n",
    "    return productions\n",
    "\n",
    "demo_bnf = \"\"\"\n",
    "<start> ::= <expr> <op> <expr>\n",
    "<expr> ::= <term> <op> <term> | '(' <term> <op> <term> ')'\n",
    "<op> ::= '+' | '-' | '/' | '*'\n",
    "<term> ::= 'x1' | '0.5'\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "def get_terminals(productions):\n",
    "    terminals = set()\n",
    "    non_terminals = set(productions.keys())\n",
    "\n",
    "    for rhs in productions.values():\n",
    "        for rule in rhs:\n",
    "            for token in rule:\n",
    "                if token not in non_terminals:\n",
    "                    terminals.add(token)\n",
    "\n",
    "    return terminals\n",
    "\n",
    "\n",
    "def create_typed_grammar(productions_dict):\n",
    "    productions = Dict.empty(key_type=types.unicode_type, value_type=types.ListType(types.ListType(types.unicode_type)))\n",
    "\n",
    "    for nt, rhs_list in productions_dict.items():\n",
    "        #rhs_typed_list = List[types.ListType(types.unicode_type)]\n",
    "\n",
    "        rhs_typed_list = typed.List.empty_list(types.ListType(types.unicode_type))\n",
    "        for rhs in rhs_list:\n",
    "            #rhs_typed = List[types.unicode_type]\n",
    "            rhs_typed = typed.List.empty_list(types.unicode_type)\n",
    "            for symbol in rhs:\n",
    "                rhs_typed.append(symbol)\n",
    "\n",
    "            rhs_typed_list.append(rhs_typed)\n",
    "\n",
    "        productions[nt] = rhs_typed_list\n",
    "\n",
    "    return productions\n",
    "\n",
    "def create_grammar_from_bnf(bnf_file, depth=1, variables=2):\n",
    "    bnf_lines = read_bnf_file(bnf_file)\n",
    "    productions = parse_bnf(bnf_lines, depth, variables)\n",
    "    productions = create_typed_grammar(productions)\n",
    "    non_terminals = set(productions.keys())\n",
    "    terminals = get_terminals(productions)\n",
    "    return productions, non_terminals, terminals\n",
    "\n",
    "productions, non_terminals, terminals = create_grammar_from_bnf(\"variable_depth.bnf\", 5, 2)\n",
    "\n",
    "pretty_print_dict(productions)"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:10:17.753922Z",
     "end_time": "2023-05-11T15:10:40.884109Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 104
    },
    "id": "yL0f36s8lCBK",
    "outputId": "c9879b84-1da8-420d-e8a8-73c853ae6e3c"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminals\n",
      "{'2', ')', '9', \"'+'\", '3', '4', '(', \"'/'\", \"'abs'\", 'x', 'y', '1', '7', '0', '8', \"'-'\", \"'*'\", '6', '5'}\n",
      "non_terminals\n",
      "{'<expr3>', '<integer>', '<non-zero-digit>', '<expr4>', '<var>', '<op>', '<digit>', '<expr>', '<start>', '<number>', '<expr2>', '<expr1>', '<uop>'}\n"
     ]
    }
   ],
   "source": [
    "terminals = get_terminals(productions)\n",
    "\n",
    "print(\"terminals\")\n",
    "print(terminals)\n",
    "non_terminals = set(productions.keys())\n",
    "print(\"non_terminals\")\n",
    "print(non_terminals)\n"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:10:40.885990Z",
     "end_time": "2023-05-11T15:10:40.892331Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 696
    },
    "id": "PrxOP6hulCBM",
    "outputId": "3a9bb23e-7be3-47df-efb6-05165e10b85a"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "data": {
      "text/plain": "'Recursive terminals:'"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": "set()"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": "'Non-recursive terminals:'"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": "{\"'*'\",\n \"'+'\",\n \"'-'\",\n \"'/'\",\n \"'abs'\",\n '(',\n ')',\n '0',\n '1',\n '2',\n '3',\n '4',\n '5',\n '6',\n '7',\n '8',\n '9',\n '<digit>',\n '<expr1>',\n '<expr2>',\n '<expr3>',\n '<expr4>',\n '<expr>',\n '<integer>',\n '<non-zero-digit>',\n '<number>',\n '<op>',\n '<start>',\n '<uop>',\n '<var>',\n 'x',\n 'y'}"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def find_recursive_and_non_recursive_terminals(grammar):\n",
    "    recursive_terminals = set()\n",
    "    non_recursive_terminals = set()\n",
    "\n",
    "    def is_recursive(nt, visited):\n",
    "        if nt in visited:\n",
    "            return True\n",
    "        visited.add(nt)\n",
    "        for rule in grammar[nt]:\n",
    "            for token in rule:\n",
    "                if token in non_terminals and is_recursive(token, visited):\n",
    "                    return True\n",
    "        visited.remove(nt)\n",
    "        return False\n",
    "\n",
    "    for nt in non_terminals:\n",
    "        if is_recursive(nt, set()):\n",
    "            recursive_terminals.add(nt)\n",
    "        else:\n",
    "            non_recursive_terminals.add(nt)\n",
    "\n",
    "    non_recursive_terminals |= terminals\n",
    "\n",
    "    return recursive_terminals, non_recursive_terminals\n",
    "\n",
    "\n",
    "recursive_terminals, non_recursive_terminals = find_recursive_and_non_recursive_terminals(productions)\n",
    "\n",
    "display(\"Recursive terminals:\", recursive_terminals)\n",
    "display(\"Non-recursive terminals:\", non_recursive_terminals)\n"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:10:40.898616Z",
     "end_time": "2023-05-11T15:10:41.647646Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 384
    },
    "id": "ECP0jRPWlCBO",
    "outputId": "679d440a-ecd8-4e00-97fd-9d3ebfd8b195"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non recursive expansions per terminal:\n",
      "<expr3>: []\n",
      "<integer>: []\n",
      "<non-zero-digit>: []\n",
      "<expr4>: []\n",
      "<var>: []\n",
      "<op>: []\n",
      "<digit>: []\n",
      "<expr>: []\n",
      "<start>: []\n",
      "<number>: []\n",
      "<expr2>: []\n",
      "<expr1>: []\n",
      "<uop>: []\n",
      "Recursive expansions per terminal:\n",
      "<expr3>: [0, 1, 2, 3]\n",
      "<integer>: [0, 1, 2]\n",
      "<non-zero-digit>: [0, 1, 2, 3, 4, 5, 6, 7, 8]\n",
      "<expr4>: [0, 1]\n",
      "<var>: [0, 1]\n",
      "<op>: [0, 1, 2, 3]\n",
      "<digit>: [0, 1]\n",
      "<expr>: [0, 1, 2, 3]\n",
      "<start>: [0]\n",
      "<number>: [0]\n",
      "<expr2>: [0, 1, 2, 3]\n",
      "<expr1>: [0, 1, 2, 3]\n",
      "<uop>: [0]\n"
     ]
    }
   ],
   "source": [
    "def calculate_non_recursive_productions(symbol, grammar):\n",
    "    non_recursive_indices = []\n",
    "    recursive_terminals_in_grammar, _ = find_recursive_and_non_recursive_terminals(grammar)\n",
    "\n",
    "    if symbol not in grammar:\n",
    "        return non_recursive_indices\n",
    "\n",
    "    for i, rule in enumerate(grammar[symbol]):\n",
    "        if all(token in recursive_terminals_in_grammar for token in rule):\n",
    "            non_recursive_indices.append(i)\n",
    "\n",
    "    return non_recursive_indices\n",
    "\n",
    "def calculate_recursive_productions(non_terminal, grammar, non_recursive_productions):\n",
    "    all_productions = set(range(len(grammar[non_terminal])))\n",
    "    non_recursive_set = set(non_recursive_productions[non_terminal])\n",
    "    return list(all_productions - non_recursive_set)\n",
    "\n",
    "def get_non_recursive_expansions(grammar):\n",
    "    non_recursive_expansions_set = OrderedDict()\n",
    "    for nt in non_terminals:\n",
    "        non_recursive_expansions_set[nt] = calculate_non_recursive_productions(nt, grammar)\n",
    "    return non_recursive_expansions_set\n",
    "\n",
    "def get_recursive_expansions(grammar, non_recursive_expansions_dict):\n",
    "    recursive_expansions_set = OrderedDict()\n",
    "    for nt in non_terminals:\n",
    "        recursive_expansions_set[nt] = calculate_recursive_productions(nt, grammar, non_recursive_expansions_dict)\n",
    "    return recursive_expansions_set\n",
    "\n",
    "# create the non-recursive dictionary for each non-terminal\n",
    "non_recursive_expansions = get_non_recursive_expansions(productions)\n",
    "recursive_expansions = get_recursive_expansions(productions, non_recursive_expansions)\n",
    "\n",
    "print(\"Non recursive expansions per terminal:\")\n",
    "pretty_print_dict(non_recursive_expansions)\n",
    "\n",
    "\n",
    "print(\"Recursive expansions per terminal:\")\n",
    "pretty_print_dict(recursive_expansions)\n"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:10:41.647646Z",
     "end_time": "2023-05-11T15:10:53.904257Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 262
    },
    "id": "C74cYPi1lCBQ",
    "outputId": "12395388-647c-4998-bc6c-a84d97629a9f"
   }
  },
  {
   "cell_type": "code",
   "source": [
    "\n",
    "\n",
    "count_references_type = Dict.empty(key_type=types.unicode_type, value_type=types.DictType(types.unicode_type, types.int64))\n",
    "is_referenced_by_type = Dict.empty(key_type=types.unicode_type, value_type=types.ListType(types.unicode_type))\n",
    "def calculate_non_terminal_references(grammar, non_terminals_set):\n",
    "    #count_references = {nt: {} for nt in non_terminals_set}\n",
    "    #count_references = Dict.empty(key_type=types.unicode_type, value_type=types.DictType(types.unicode_type, types.int64))\n",
    "    \"\"\"\n",
    "    count_references = dict()\n",
    "    for nt in non_terminals_set:\n",
    "        #count_references[nt] = Dict.empty(key_type=types.unicode_type, value_type=types.int64)\n",
    "        count_references[nt] = dict()\n",
    "    #is_referenced_by = Dict.empty(key_type=types.unicode_type, value_type=types.unicode_type[:])\n",
    "    is_referenced_by = dict()\n",
    "    for nt in non_terminals_set:\n",
    "        #is_referenced_by[nt] = List.empty_list(types.int64)\n",
    "        is_referenced_by[nt] = []\n",
    "    #is_referenced_by = {nt: [] for nt in non_terminals_set}\n",
    "    \"\"\"\n",
    "    count_references = count_references_type.copy()\n",
    "    is_referenced_by = is_referenced_by_type.copy()\n",
    "\n",
    "    for nt in non_terminals_set:\n",
    "        count_references[nt] = Dict.empty(key_type=types.unicode_type, value_type=types.int64)\n",
    "        is_referenced_by[nt] = List.empty_list(types.unicode_type)\n",
    "\n",
    "    for nt in non_terminals_set:\n",
    "        for production in grammar[nt]:\n",
    "            count = {option: 0 for option in non_terminals_set}\n",
    "            for option in production:\n",
    "                if option in non_terminals_set:\n",
    "                    is_referenced_by[option].append(nt)\n",
    "                    count[option] += 1\n",
    "            for key in count:\n",
    "                count_references[key][nt] = max(count_references[key].get(nt, 0), count[key])\n",
    "\n",
    "    return count_references, is_referenced_by\n",
    "\n",
    "# Usage example\n",
    "count_refs, ref_by = calculate_non_terminal_references(productions, non_terminals)\n",
    "print(\"count_refs\")\n",
    "\n",
    "pretty_print_dict(count_refs)\n",
    "print(\"Ref by: \")\n",
    "pretty_print_dict(ref_by)"
   ],
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "vzXV0Gxy4rry",
    "outputId": "f83b73bc-db8e-4c21-aebd-cba5f9cfc68c",
    "ExecuteTime": {
     "start_time": "2023-05-11T15:10:53.910609Z",
     "end_time": "2023-05-11T15:11:09.708778Z"
    }
   },
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count_refs\n",
      "<expr3>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 2, <expr1>: 0, <uop>: 0}\n",
      "<integer>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 1, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<non-zero-digit>: {<expr3>: 0, <integer>: 1, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 1, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<expr4>: {<expr3>: 2, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<var>: {<expr3>: 1, <integer>: 0, <non-zero-digit>: 0, <expr4>: 1, <var>: 0, <op>: 0, <digit>: 0, <expr>: 1, <start>: 0, <number>: 0, <expr2>: 1, <expr1>: 1, <uop>: 0}\n",
      "<op>: {<expr3>: 1, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 1, <start>: 0, <number>: 0, <expr2>: 1, <expr1>: 1, <uop>: 0}\n",
      "<digit>: {<expr3>: 0, <integer>: 2, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<expr>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 1, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<start>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<number>: {<expr3>: 1, <integer>: 0, <non-zero-digit>: 0, <expr4>: 1, <var>: 0, <op>: 0, <digit>: 0, <expr>: 1, <start>: 0, <number>: 0, <expr2>: 1, <expr1>: 1, <uop>: 0}\n",
      "<expr2>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 0, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 2, <uop>: 0}\n",
      "<expr1>: {<expr3>: 0, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 2, <start>: 0, <number>: 0, <expr2>: 0, <expr1>: 0, <uop>: 0}\n",
      "<uop>: {<expr3>: 1, <integer>: 0, <non-zero-digit>: 0, <expr4>: 0, <var>: 0, <op>: 0, <digit>: 0, <expr>: 1, <start>: 0, <number>: 0, <expr2>: 1, <expr1>: 1, <uop>: 0}\n",
      "Ref by: \n",
      "<expr3>: [<expr2>, <expr2>, <expr2>, ...]\n",
      "<integer>: [<number>, ...]\n",
      "<non-zero-digit>: [<integer>, <integer>, <digit>, ...]\n",
      "<expr4>: [<expr3>, <expr3>, <expr3>, ...]\n",
      "<var>: [<expr3>, <expr4>, <expr>, <expr2>, <expr1>, ...]\n",
      "<op>: [<expr3>, <expr>, <expr2>, <expr1>, ...]\n",
      "<digit>: [<integer>, <integer>, <integer>, <integer>, ...]\n",
      "<expr>: [<start>, ...]\n",
      "<start>: [, ...]\n",
      "<number>: [<expr3>, <expr4>, <expr>, <expr2>, <expr1>, ...]\n",
      "<expr2>: [<expr1>, <expr1>, <expr1>, ...]\n",
      "<expr1>: [<expr>, <expr>, <expr>, ...]\n",
      "<uop>: [<expr3>, <expr>, <expr2>, <expr1>, ...]\n"
     ]
    }
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "data": {
      "text/plain": "320"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@jit(nopython=True)\n",
    "def find_references(nt, start_symbol, is_referenced_by, count_references_by_prod):\n",
    "    r = get_total_references_of_current_production(count_references_by_prod, nt)\n",
    "    results = []\n",
    "\n",
    "    if nt == start_symbol:\n",
    "        return 1\n",
    "    nt_str = str(nt)\n",
    "    for ref in is_referenced_by[nt_str]:\n",
    "        results.append(find_references(ref, start_symbol, is_referenced_by, count_references_by_prod))\n",
    "\n",
    "    references = r * np.max(np.array(results))\n",
    "    return references\n",
    "\n",
    "@jit(nopython=True)\n",
    "def get_total_references_of_current_production(count_references_by_prod, nt):\n",
    "    nt_str = str(nt)\n",
    "    return np.sum(np.array(list(count_references_by_prod[nt_str].values())))\n",
    "\n",
    "\n",
    "nt = \"<digit>\"  # Replace this with a non-terminal from your grammar\n",
    "nt = '<non-zero-digit>'\n",
    "references = find_references(nt, '<start>', ref_by, count_refs)\n",
    "\n",
    "display(references)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:11:09.736096Z",
     "end_time": "2023-05-11T15:11:16.871593Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "data": {
      "text/plain": "'\\ngenotype, grammar = create_genotype(max_depth=3)\\n#pretty_print_dict(grammar)\\ntree = Tree(genotype, grammar)\\nprint(\"Tree created!\")\\n\\nwhile tree.expand_next():\\n    pass\\n\\n\\nprint(\"Done\")\\ndisplay(tree)\\n'"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Tree:\n",
    "\n",
    "    def __init__(self, genome, productions):\n",
    "        # get first rule for start symbol from the grammar\n",
    "\n",
    "        self.productions = productions\n",
    "        self.non_terminals = set(productions.keys())\n",
    "        first_rule = next(iter(productions))\n",
    "        first_production = productions[first_rule][0]\n",
    "        self.root = Node(first_rule)\n",
    "\n",
    "        # initialize OrderedDict with each non-terminal as a key and an empty list as the value, without list comprehension\n",
    "        self.genome = genome\n",
    "\n",
    "    def __repr__(self):\n",
    "        #return f\"Node({self.label}, {self.children})\"\n",
    "        # if it has children, call repr on each child\n",
    "        return self.root.children[0].simple_repr()\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.root.__str__()\n",
    "\n",
    "    def _get_next_expansion(self):\n",
    "        # find the first non-terminal that has not been expanded\n",
    "        return self.root.find_first_unexpanded_non_terminal()\n",
    "\n",
    "    def expand_next(self):\n",
    "        node = self._get_next_expansion()\n",
    "        if node is None:\n",
    "            return False\n",
    "\n",
    "        # get the vector of production indices for the current non-terminal\n",
    "        production_indices = self.genome[node.label]\n",
    "\n",
    "        if len(production_indices) == 0:\n",
    "            raise ValueError(f\"Genome for {node.label} is empty\")\n",
    "\n",
    "\n",
    "        # get the next production index\n",
    "        production_index = production_indices.pop(0)\n",
    "\n",
    "        # get the production for the desired non-terminal\n",
    "        new_production = self.productions[node.label][production_index]\n",
    "\n",
    "        #print(\"Expanding\", node.label, \"with\", new_production)\n",
    "        node.apply_rule(self.productions, new_production)\n",
    "        return True\n",
    "\n",
    "\n",
    "\n",
    "class Node:\n",
    "\n",
    "    def __init__(self, label, first_production=None):\n",
    "        #print(\"Creating node with label\", label)\n",
    "        self.label = label\n",
    "        self.children = []\n",
    "        self.is_terminal = False\n",
    "\n",
    "        if first_production is not None and len(first_production) > 0:\n",
    "            self.apply_rule(first_production)\n",
    "\n",
    "    def simple_repr(self):\n",
    "        if len(self.children) > 0:\n",
    "            merged_string = \" \".join([child.simple_repr() for child in self.children])\n",
    "            return merged_string\n",
    "        return self.label\n",
    "\n",
    "    def find_first_unexpanded_non_terminal(self):\n",
    "        if self.is_terminal:\n",
    "            return None\n",
    "\n",
    "        if len(self.children) == 0:\n",
    "            return self\n",
    "\n",
    "        for child in self.children:\n",
    "            non_terminal = child.find_first_unexpanded_non_terminal()\n",
    "            if non_terminal is not None:\n",
    "                return non_terminal\n",
    "\n",
    "        return None\n",
    "\n",
    "    def apply_rule(self, current_tree_productions, production: list):\n",
    "\n",
    "        if production not in current_tree_productions[self.label]:\n",
    "            raise ValueError(f\"Production {production} not found in grammar for {self.label}\")\n",
    "\n",
    "        children_list = []\n",
    "        for symbol in production:\n",
    "            child = Node(symbol)\n",
    "            if child.label not in current_tree_productions:\n",
    "                child.is_terminal = True\n",
    "            children_list.append(child)\n",
    "\n",
    "        self.children = children_list\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "genotype, grammar = create_genotype(max_depth=3)\n",
    "#pretty_print_dict(grammar)\n",
    "tree = Tree(genotype, grammar)\n",
    "print(\"Tree created!\")\n",
    "\n",
    "while tree.expand_next():\n",
    "    pass\n",
    "\n",
    "\n",
    "print(\"Done\")\n",
    "display(tree)\n",
    "\"\"\""
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:11:16.902893Z",
     "end_time": "2023-05-11T15:11:17.093042Z"
    },
    "id": "EJboc1OYlCBT"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [
    {
     "data": {
      "text/plain": "(DictType[unicode_type,ListType[int64]]<iv=None>({<integer>: [2, 0, 1, 0, 0, 1, 0, 1, 0, 2, 2, 0, ...], <non-zero-digit>: [6, 7, 6, 6, 4, 0, 1, 5, 8, 6, 6, 1, 0, 1, 1, 2, 7, 8, 1, 7, 4, 1, 6, 0, 2, 5, 3, 3, 4, 6, 3, 2, 0, 1, 5, 4, 8, 6, 3, 0, 8, 7, 5, 7, 4, 5, 0, 7, ...], <var>: [0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, ...], <op>: [2, 2, 1, 3, ...], <digit>: [1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, ...], <expr>: [3, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...], <expr2>: [1, 1, 0, 1, ...], <expr1>: [0, 1, ...], <uop>: [0, 0, 0, 0, ...]}),\n DictType[unicode_type,ListType[ListType[unicode_type]]]<iv=None>({<start>: [[<expr>, ...], ...], <expr>: [[<number>, ...], [<var>, ...], [(, <op>, <expr1>, <expr1>, ), ...], [(, <uop>, <expr1>, ), ...], ...], <number>: [[<integer>, ...], ...], <integer>: [[(, <digit>, ), ...], [(, <non-zero-digit>, <digit>, ), ...], [(, <non-zero-digit>, <digit>, <digit>, ), ...], ...], <digit>: [[0, ...], [<non-zero-digit>, ...], ...], <non-zero-digit>: [[1, ...], [2, ...], [3, ...], [4, ...], [5, ...], [6, ...], [7, ...], [8, ...], [9, ...], ...], <op>: [['+', ...], ['-', ...], ['*', ...], ['/', ...], ...], <uop>: [['abs', ...], ...], <var>: [[x, ...], [y, ...], ...], <expr1>: [[<number>, ...], [<var>, ...], [(, <op>, <expr2>, <expr2>, ), ...], [(, <uop>, <expr2>, ), ...], ...], <expr2>: [[<number>, ...], [<var>, ...], ...]}))"
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@njit\n",
    "def seed(a):\n",
    "    random.seed(a)\n",
    "\n",
    "@njit\n",
    "def rand():\n",
    "    return random.random()\n",
    "\n",
    "@njit\n",
    "def random_int(a, b):\n",
    "    return random.randint(a, b)\n",
    "\n",
    "\n",
    "production_type = types.ListType(types.int64)\n",
    "new_genotype_type = Dict.empty(\n",
    "    key_type=types.unicode_type,\n",
    "    value_type=production_type\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "@jit(nopython=False)\n",
    "def create_full_tree(grammar, genotype,first_symbol, ref_by_dict, count_refs_dict, non_terminals_set):\n",
    "    #print(non_terminals_set)\n",
    "    #print(numba.typeof(non_terminals_set))\n",
    "    #print(numba.version_info)\n",
    "    # for every non-terminal, we create a vector of size equal to the upper bound of productions\n",
    "    non_terminals_list = list(non_terminals_set)\n",
    "    for symbol_index in prange(len(non_terminals_list)):\n",
    "        symbol = str(non_terminals_list[symbol_index])\n",
    "\n",
    "        upper_bound = find_references(symbol, first_symbol, ref_by_dict, count_refs_dict)\n",
    "\n",
    "        productions_length = len(grammar[symbol])\n",
    "        possible_productions = List.empty_list(types.int64)\n",
    "        for i in range(upper_bound):\n",
    "            codon = random_int(0, productions_length - 1)\n",
    "            possible_productions.append(codon)\n",
    "\n",
    "        genotype[symbol] = possible_productions\n",
    "\n",
    "def create_individual_probabilistic(grammar, max_depth, genotype, symbol, non_terminals_set, depth):\n",
    "    stack = [(symbol, depth)]\n",
    "\n",
    "    is_terminal_cache = {s not in non_terminals_set for s in grammar}\n",
    "    unique_depths = set()\n",
    "    non_recursive_expansions_dict = get_non_recursive_expansions(grammar)\n",
    "\n",
    "    while stack:\n",
    "        symbol, depth = stack.pop()\n",
    "        if depth not in unique_depths:\n",
    "          #print(f\"Reached new unique depth: {depth}\")\n",
    "          unique_depths.add(depth)\n",
    "        production_rules = grammar[symbol]\n",
    "        expansion_index = random.randint(0, len(production_rules) - 1)\n",
    "\n",
    "        expansion = production_rules[expansion_index]\n",
    "\n",
    "        # handle case where symbol is recursive, checking the dict\n",
    "        is_expansion_rec = expansion in recursive_expansions[symbol]\n",
    "        #if not is_symbol_rec:\n",
    "          #print(f\"Symbol {symbol} is non recursive\")\n",
    "        if is_expansion_rec:\n",
    "              if depth >= max_depth:\n",
    "                  non_rec_exps = non_recursive_expansions_dict[symbol]\n",
    "                  if len(non_rec_exps) == 0:\n",
    "                      print(\"Symbol\", symbol, \"has no non-recursive productions\")\n",
    "                      raise ValueError(\"No valid productions in this case\")\n",
    "                  expansion_index = random.choice(non_rec_exps)\n",
    "                  expansion = grammar[symbol][expansion_index]\n",
    "\n",
    "        if symbol in genotype:\n",
    "            genotype[symbol].append(expansion_index)\n",
    "        else:\n",
    "            genotype[symbol] = [expansion_index]\n",
    "\n",
    "        expansion_symbols = production_rules[expansion_index]\n",
    "\n",
    "        for sym in expansion_symbols:\n",
    "            if not is_terminal_cache.get(sym, True):\n",
    "                stack.append((sym, depth + 1))\n",
    "    print(\"Unique depths: \", unique_depths)\n",
    "#\n",
    "# def create_individual_recursive(grammar, max_depth, genotype, symbol, depth):\n",
    "#     production_rules = grammar[symbol]\n",
    "#     expansion_index = random.randint(0, len(production_rules) - 1)\n",
    "#\n",
    "#     expansion = production_rules[expansion_index]\n",
    "#     # check if the symbol is a recursive terminal, ie, it can expand to itself\n",
    "#     if is_recursive(symbol):\n",
    "#         # check if the expansion is recursive\n",
    "#         if expansion in recursive_expansions[symbol]:\n",
    "#             if depth >= max_depth:\n",
    "#                 # get non recursive productions of the symbol\n",
    "#                 non_rec_exps = non_recursive_expansions[symbol]\n",
    "#                 if len(non_rec_exps) == 0:\n",
    "#                     print(\"Symbol\", symbol, \"has no non-recursive productions\")\n",
    "#                     return\n",
    "#                 expansion_index = random.choice(non_rec_exps)\n",
    "#                 expansion = grammar[symbol][expansion_index]\n",
    "#     else:\n",
    "#       print(f\"Symbol {symbol} is non recursive!\")\n",
    "#\n",
    "#     if symbol in genotype:\n",
    "#         genotype[symbol].append(expansion_index)\n",
    "#     else:\n",
    "#         genotype[symbol] = [expansion_index]\n",
    "#\n",
    "#     expansion_symbols = production_rules[expansion_index]\n",
    "#\n",
    "#     for sym in expansion_symbols:\n",
    "#         if not is_terminal(sym):\n",
    "#             create_individual_probabilistic(grammar, max_depth, genotype, sym, depth + 1)\n",
    "\n",
    "def create_genotype(grammar_file='variable_depth.bnf', max_depth=6, variables_count=2, option='full'):\n",
    "    new_genotype = new_genotype_type.copy()\n",
    "\n",
    "    desired_depth_grammar, non_terminals, terminals = create_grammar_from_bnf(grammar_file, max_depth, variables_count)\n",
    "    first_symbol = next(iter(desired_depth_grammar.keys()))\n",
    "\n",
    "    count_refs, ref_by = calculate_non_terminal_references(desired_depth_grammar, non_terminals)\n",
    "\n",
    "    if option == 'full':\n",
    "        create_full_tree(desired_depth_grammar, new_genotype, first_symbol, ref_by, count_refs, non_terminals)\n",
    "    elif option == 'probabilistic':\n",
    "        create_individual_probabilistic(desired_depth_grammar, max_depth, new_genotype, first_symbol, 0)\n",
    "    return new_genotype, desired_depth_grammar\n",
    "\n",
    "\n",
    "create_genotype(max_depth=3)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:11:16.952172Z",
     "end_time": "2023-05-11T15:11:32.081951Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "data": {
      "text/plain": "\"( '/' ( 0 ) x )\""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": "( '/' ( 0 ) x )"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# NodeData class definition\n",
    "node_data_spec = [\n",
    "    ('id', types.int64),\n",
    "    ('label', types.unicode_type),\n",
    "    ('children', types.ListType(types.int64)),\n",
    "    ('is_terminal', types.boolean),\n",
    "]\n",
    "\n",
    "@jitclass(node_data_spec)\n",
    "class NodeData:\n",
    "    def __init__(self, node_id, label):\n",
    "        self.id = node_id\n",
    "        self.label = label\n",
    "        self.children = List.empty_list(types.int64)\n",
    "        self.is_terminal = False\n",
    "\n",
    "# Global hash table to store NodeData instances\n",
    "nodes = Dict.empty(key_type=types.int64, value_type=NodeData.class_type.instance_type)\n",
    "\n",
    "# Function to create a new NodeData instance and store it in the global hash table\n",
    "@njit\n",
    "def create_node(label, nodes_table):\n",
    "    id = len(nodes_table)\n",
    "    node = NodeData(id, label)\n",
    "    nodes_table[id] = node\n",
    "    return id\n",
    "\n",
    "# Functions to work with NodeData instances\n",
    "@njit\n",
    "def apply_rule(node_id, node_table, grammar, production):\n",
    "    node = node_table[node_id]\n",
    "\n",
    "    found = False\n",
    "\n",
    "    # print current productions\n",
    "    #print(\"Current productions for\", node.label, \"are\", current_tree_productions[node.label])\n",
    "\n",
    "    for prods in grammar[node.label]:\n",
    "        #print(set(production), set(prods))\n",
    "        if set(production) == set(prods):\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        raise ValueError(f\"Production {production} not found in grammar for {node.label}\")\n",
    "\n",
    "    #if production not in current_tree_productions[node.label]:\n",
    "    #    raise ValueError(f\"Production {production} not found in grammar for {node.label}\")\n",
    "\n",
    "    children_list = List.empty_list(types.int64)\n",
    "    for symbol in production:\n",
    "        child_id = create_node(symbol, node_table)\n",
    "        current_node = node_table[child_id]\n",
    "\n",
    "        found = False\n",
    "        for prod in grammar:\n",
    "            #print(prod)\n",
    "            if current_node.label == prod:\n",
    "                found = True\n",
    "                #print(f\"Found {current_node.label} in {prod}, not terminal\")\n",
    "                break\n",
    "        if not found:\n",
    "            current_node.is_terminal = True\n",
    "\n",
    "        children_list.append(child_id)\n",
    "\n",
    "    node.children = children_list\n",
    "\n",
    "@njit\n",
    "def find_first_unexpanded_non_terminal(node_id: int, node_table, productions):\n",
    "    node = node_table[node_id]\n",
    "    if node.is_terminal:\n",
    "        return -1\n",
    "\n",
    "    if len(node.children) == 0:\n",
    "        return node_id\n",
    "\n",
    "    for child_id in node.children:\n",
    "        non_terminal_idx = find_first_unexpanded_non_terminal(child_id, node_table, productions)\n",
    "        if non_terminal_idx != -1:\n",
    "            return non_terminal_idx\n",
    "\n",
    "    return -1\n",
    "    # Implement find_first_unexpanded_non_terminal functionality\n",
    "\n",
    "@njit\n",
    "def expand_next(node_id, node_table, genome, productions):\n",
    "    expansion_node_id = find_first_unexpanded_non_terminal(node_id, node_table, productions)\n",
    "\n",
    "    if expansion_node_id == -1:\n",
    "        return False\n",
    "\n",
    "    # get the vector of production indices for the current non-terminal\n",
    "    node = node_table[expansion_node_id]\n",
    "    production_indices = genome[node.label]\n",
    "\n",
    "    #print(f\"Node: {node.label}, children = {node.children}, genome={genome}\")\n",
    "\n",
    "    if len(production_indices) == 0:\n",
    "        print(f\"Node: {node.label}, children = {node.children}, genome={genome}\")\n",
    "        raise ValueError(f\"Genome for {node.label} is empty\")\n",
    "\n",
    "    # get the next production index\n",
    "    production_index = production_indices.pop(0)\n",
    "\n",
    "    # get the production for the desired non-terminal\n",
    "    new_production = productions[node.label][production_index]\n",
    "\n",
    "    apply_rule(expansion_node_id, node_table, productions, new_production)\n",
    "    return True\n",
    "\n",
    "@njit\n",
    "def load_genotype(node_table, genome, productions, print_tree=False, print_aplications=False):\n",
    "    counter = 0\n",
    "    while expand_next(0, node_table, genome, productions):\n",
    "        counter += 1\n",
    "\n",
    "    if print_aplications:\n",
    "        print(\"Applied rule \", counter, \" times\")\n",
    "        print_nodes(node_table)\n",
    "        print()\n",
    "    if print_tree:\n",
    "        simple_repr(0, node_table)\n",
    "\n",
    "\n",
    "@njit\n",
    "def simple_repr(node_id, node_table):\n",
    "    node = node_table[node_id]\n",
    "\n",
    "    if len(node.children) > 0:\n",
    "        merged_string = \"\"\n",
    "        for child_id in node.children:\n",
    "            merged_string += simple_repr(child_id, node_table) + \" \"\n",
    "        return merged_string.strip()\n",
    "    return node.label\n",
    "\n",
    "@njit\n",
    "def print_nodes(nodes_table):\n",
    "    for node_id in nodes_table.keys():\n",
    "        node = nodes_table[node_id]\n",
    "        print(f\"{node_id}: {node.label}\")\n",
    "\n",
    "def generate_and_expand(node_table, depth=4, variables_count=2):\n",
    "    genotype, grammar = create_genotype(max_depth=depth, variables_count=variables_count)\n",
    "    copy_genome = {}\n",
    "\n",
    "    for key in genotype:\n",
    "        copy_genome[key] = genotype[key].copy()\n",
    "\n",
    "    tree = Tree(copy_genome, grammar)\n",
    "    while tree.expand_next():\n",
    "        pass\n",
    "    load_genotype(node_table, genotype, grammar)\n",
    "    display(simple_repr(0, node_table))\n",
    "    display(tree)\n",
    "    return tree\n",
    "\n",
    "# Other functions as needed\n",
    "\n",
    "create_node('<start>', nodes)\n",
    "\"\"\"\n",
    "apply_rule(0, nodes, grammar, ['<expr>'])\n",
    "second_production = grammar['<expr>'][2]\n",
    "apply_rule(1, nodes, grammar, second_production)\n",
    "third_production = grammar['<op>'][0]\n",
    "apply_rule(3, nodes, grammar, third_production)\n",
    "fourth_production = grammar['<expr1>'][1]\n",
    "apply_rule(4, nodes, grammar, fourth_production)\n",
    "apply_rule(5, nodes, grammar, fourth_production)\n",
    "\"\"\"\n",
    "tree = generate_and_expand(nodes, depth=6, variables_count=2)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:11:32.137307Z",
     "end_time": "2023-05-11T15:12:11.359011Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "( '/' ( 0 ) x )\n"
     ]
    },
    {
     "data": {
      "text/plain": "0.0"
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from math import nan\n",
    "\n",
    "@njit\n",
    "def remove_quotes(target: str) -> str:\n",
    "    # replace quotes with nothing\n",
    "    #return target.replace(\"'\", \"\").replace('\"', '')\n",
    "    result = \"\"\n",
    "    for char in target:\n",
    "        if char != \"'\" and char != '\"':\n",
    "            result += char\n",
    "    return result\n",
    "\n",
    "@njit\n",
    "def convertStrToInt(s: str) -> int:\n",
    "    sign = 1\n",
    "    if s.startswith('-'):\n",
    "        sign = -1\n",
    "        s = s[1:]\n",
    "    elif s.startswith('+'):\n",
    "        s = s[1:]\n",
    "\n",
    "    integer_value = 0\n",
    "    for c in s:\n",
    "        if c < '0' or c > '9':\n",
    "            return 0  # or you can return some error code to indicate the conversion failed\n",
    "\n",
    "        digit = ord(c) - ord('0')\n",
    "        integer_value = integer_value * 10 + digit\n",
    "\n",
    "    return sign * integer_value\n",
    "@njit\n",
    "def evaluate_string(node_id: int, node_table):\n",
    "    node = node_table[node_id]\n",
    "    return node.label\n",
    "\n",
    "\n",
    "@njit\n",
    "def evaluate(node_id: int, node_table, variables_dict) -> float:\n",
    "    node = node_table[node_id]\n",
    "    label = node.label\n",
    "\n",
    "    #print(\"Evaluating \", label)\n",
    "\n",
    "    if label == \"<start>\":\n",
    "        return evaluate(node.children[0], node_table, variables_dict)\n",
    "    elif node.label.startswith('<expr'):\n",
    "        if len(node.children) == 1:  # <number>\n",
    "            return evaluate(node.children[0], node_table, variables_dict)\n",
    "        else:\n",
    "            operator_wrapper_node_id = node.children[1]\n",
    "            operator_node_id = node_table[operator_wrapper_node_id].children[0]\n",
    "            if len(node.children) == 4:  # ( <uop> <expr> )\n",
    "                uop = evaluate_string(operator_node_id, node_table)\n",
    "                uop = remove_quotes(uop)\n",
    "                operand_node_id = node.children[2]\n",
    "                operand = evaluate(operand_node_id, node_table, variables_dict)\n",
    "\n",
    "                if uop == 'abs':\n",
    "                    return abs(operand)\n",
    "                else:\n",
    "                    print(\"Error, unknown unary operator\")\n",
    "                    return nan\n",
    "            elif len(node.children) == 5:  # ( <op> <expr> <expr> )\n",
    "\n",
    "                op = evaluate_string(operator_node_id, node_table)\n",
    "                op = remove_quotes(op)\n",
    "                op1 = evaluate(node.children[2], node_table, variables_dict)\n",
    "                op2 = evaluate(node.children[3], node_table, variables_dict)\n",
    "                if op == \"+\":\n",
    "                    return op1 + op2\n",
    "                elif op == \"-\":\n",
    "                    return op1 - op2\n",
    "                elif op == \"*\":\n",
    "                    return op1 * op2\n",
    "                elif op == \"/\":\n",
    "                    # safe division\n",
    "                    if op2 == 0:\n",
    "                        return 0;\n",
    "                    return op1 / op2\n",
    "                else:\n",
    "                    print(\"Error, unknown binary operator: \", op)\n",
    "                    return nan\n",
    "\n",
    "    elif label == \"<number>\":\n",
    "        return evaluate(node.children[0], node_table, variables_dict)\n",
    "\n",
    "    elif label == \"<integer>\":\n",
    "        # Remove nodes where label is parenthesis from children\n",
    "        digit_node_ids = List.empty_list(int64)\n",
    "        for child_id in node.children:\n",
    "            curr = node_table[child_id]\n",
    "            if curr.label != '(' and curr.label != ')':\n",
    "                #print(\"Adding \", curr.label, \" to digit_node_ids, with id\", child_id)\n",
    "                digit_node_ids.append(child_id)\n",
    "\n",
    "        # Evaluate each digit node and join them together\n",
    "        merged_digits = \"\"\n",
    "        for child_id in digit_node_ids:\n",
    "            number_id = node_table[child_id].children[0]\n",
    "            # evaluate returns a float, we need to cast to int\n",
    "            evaluated_digit = int(evaluate(number_id, node_table, variables_dict))\n",
    "            digit_value = str(evaluated_digit)\n",
    "            #print(\"Digit value is \", digit_value)\n",
    "            merged_digits += digit_value\n",
    "        temp = convertStrToInt(merged_digits)\n",
    "        return float(temp)\n",
    "\n",
    "    elif label == \"<var>\":\n",
    "        desired_variable = evaluate_string(node.children[0], node_table)\n",
    "        desired_variable = remove_quotes(desired_variable)\n",
    "\n",
    "        if desired_variable not in variables_dict:\n",
    "            raise ValueError(f\"Variable {desired_variable} not found in variables {variables_dict}\")\n",
    "        return float(variables_dict[desired_variable])\n",
    "\n",
    "    elif label in [\"<non-zero-digit>\", \"<digit>\"]:\n",
    "        return evaluate(node.children[0], node_table, variables_dict)\n",
    "    elif label in \"0123456789\":\n",
    "        #print(\"Digit found! \", label)\n",
    "        temp = convertStrToInt(label)\n",
    "        #print(\"Converted to \", temp)\n",
    "        return float(temp)\n",
    "\n",
    "    print(f\"Unexpected node label: {label}\")\n",
    "    return nan\n",
    "\n",
    "variables = Dict.empty(key_type=types.unicode_type, value_type=types.float64)\n",
    "\n",
    "# Add variables to the dictionary\n",
    "variables['x'] = 42.0\n",
    "variables['y'] = 10.0\n",
    "#print_nodes(nodes)\n",
    "print(simple_repr(0, nodes))\n",
    "evaluate(0, nodes, variables)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:12:11.387558Z",
     "end_time": "2023-05-11T15:12:42.489187Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "def plot_tree(tree):\n",
    "    G = nx.DiGraph()\n",
    "\n",
    "    def add_edges(node):\n",
    "        for child in node.children:\n",
    "            G.add_edge(node.label, child.label)\n",
    "            add_edges(child)\n",
    "\n",
    "    add_edges(tree.root)\n",
    "\n",
    "    pos = nx.drawing.nx_agraph.graphviz_layout(G, prog=\"dot\")\n",
    "    nx.draw(G, pos, with_labels=True, node_size=2000, node_color=\"skyblue\", font_size=12, font_weight=\"bold\", arrowsize=20)\n",
    "    plt.show()\n",
    "\n"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:12:42.498630Z",
     "end_time": "2023-05-11T15:12:42.503643Z"
    },
    "id": "cZLcMon-lCBW"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [],
   "source": [
    "def clean_string(target: str):\n",
    "    # replace quotes with nothing\n",
    "    return target.replace(\"'\", \"\").replace('\"', '')\n",
    "def evaluate_slow(node: Node, variables):\n",
    "    if node.label == \"<start>\":\n",
    "        return evaluate_slow(node.children[0], variables)\n",
    "    elif re.match(r'<expr\\d*>', node.label):\n",
    "        if len(node.children) == 1:  # <number>\n",
    "            return evaluate_slow(node.children[0], variables)\n",
    "        else:\n",
    "            operator_node = node.children[1]\n",
    "            if len(node.children) == 4:  # ( <uop> <expr> )\n",
    "                #print(\"Operator node: \", operator_node)\n",
    "                uop = evaluate_slow(operator_node, variables)\n",
    "                uop = clean_string(uop)\n",
    "                #print(\"Evaluated operator: [\", uop, \"]\")\n",
    "                operand_node = node.children[2]\n",
    "                operand = evaluate_slow(operand_node, variables)\n",
    "                if operand is None:\n",
    "                    print(\"Operand was \", operand_node, \"before evaluation, error\")\n",
    "                    #operand = evaluate(operand_node)\n",
    "                if uop == 'abs':\n",
    "                    #print(\"Doing abs of \", operand, \" =\",abs(operand))\n",
    "                    return abs(operand)\n",
    "                else:\n",
    "                    raise ValueError(f\"Unknown unary operator {uop}\")\n",
    "            elif len(node.children) == 5:  # ( <op> <expr> <expr> )\n",
    "                op = evaluate_slow(operator_node, variables)\n",
    "                op = clean_string(op)\n",
    "                op1 = evaluate_slow(node.children[2], variables)\n",
    "                op2 = evaluate_slow(node.children[3], variables)\n",
    "                if op == \"+\":\n",
    "                    return op1 + op2\n",
    "                elif op == \"-\":\n",
    "                    return op1 - op2\n",
    "                elif op == \"*\":\n",
    "                    return op1 * op2\n",
    "                elif op == \"/\":\n",
    "                    # safe division\n",
    "                    if op2 == 0:\n",
    "                        return 0;\n",
    "                    return op1 / op2\n",
    "\n",
    "    elif node.label == \"<number>\":\n",
    "        return evaluate_slow(node.children[0], variables)\n",
    "\n",
    "    elif node.label == \"<integer>\":\n",
    "        # Remove nodes where label is parenthesis from children\n",
    "        parenthesis =  ['(', ')']\n",
    "        digit_nodes = [child for child in node.children if child.label not in parenthesis]\n",
    "\n",
    "        # Evaluate each digit node and join them together\n",
    "        merged_digits = [str(evaluate_slow(child, variables)) for child in digit_nodes]\n",
    "        return int(\"\".join(merged_digits))\n",
    "\n",
    "    elif node.label == \"<var>\":\n",
    "        desired_variable = node.children[0].label\n",
    "        desired_variable = clean_string(desired_variable)\n",
    "        if desired_variable not in variables:\n",
    "            raise ValueError(f\"Variable {desired_variable} not found in variables {variables}\")\n",
    "        return variables[desired_variable]\n",
    "    elif node.label == \"<op>\":\n",
    "        return node.children[0].label\n",
    "    elif node.label == \"<uop>\":\n",
    "        return node.children[0].label\n",
    "    elif node.label in [\"<non-zero-digit>\", \"<digit>\"]:\n",
    "        return evaluate_slow(node.children[0], variables)\n",
    "    elif node.label in \"0123456789\":\n",
    "        return int(node.label)\n",
    "    elif node.label == \"(\" or node.label == \")\":\n",
    "        return \"\"\n",
    "    else:\n",
    "        raise ValueError(f\"Unexpected node label: {node.label}\")\n"
   ],
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-11T15:12:42.503643Z",
     "end_time": "2023-05-11T15:12:42.583176Z"
    },
    "id": "3icHU_YslCBY"
   }
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes:  2000\n",
      "Unique genotypes:  2000\n",
      "First genotype:  {<integer>: [2, 1, 0, 1, ...], <non-zero-digit>: [6, 8, 0, 4, 2, 0, 2, 4, 6, 4, 5, 5, 8, 3, 7, 0, ...], <var>: [0, 1, 0, 0, ...], <op>: [1, ...], <digit>: [1, 1, 0, 0, 0, 1, 1, 0, ...], <expr>: [1, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, ...], <expr1>: [0, 1, ...], <uop>: [0, ...]}\n"
     ]
    }
   ],
   "source": [
    "import concurrent.futures\n",
    "\n",
    "def create_single_genotype(args):\n",
    "    max_depth, variables_count = args\n",
    "    try:\n",
    "        genotype, _ = create_genotype(max_depth=max_depth, variables_count=variables_count)\n",
    "        return genotype\n",
    "    except Exception as e:\n",
    "        print(\"Error creating genotype: \", e)\n",
    "        raise e\n",
    "\n",
    "def create_n_genotypes(n: int, max_depth: int, variables_count=3) -> List[new_genotype_type]:\n",
    "    with concurrent.futures.ThreadPoolExecutor() as executor:\n",
    "        genotypes = list(executor.map(create_single_genotype, [(max_depth, variables_count) for _ in range(n)]))\n",
    "    _, grammar = create_genotype(max_depth=max_depth, variables_count=variables_count)\n",
    "    #genotypes = [create_single_genotype((max_depth, variables_count)) for _ in range(n)]\n",
    "    #print(\"First created genotype: \", genotypes[0])\n",
    "    return genotypes, grammar\n",
    "\n",
    "def test_n_creation():\n",
    "    # Example usage\n",
    "    n = 2000\n",
    "    depth = 2\n",
    "    variables = {'x': 42, 'y': 10}\n",
    "\n",
    "    genotypes, grammar = create_n_genotypes(n, depth, len(variables.keys()))\n",
    "    assert len(genotypes) == n , \"Should create n genotypes\"\n",
    "\n",
    "    all_hashes = [genotype_hash(genotype) for genotype in genotypes]\n",
    "    print(\"Genotypes: \", len(genotypes))\n",
    "    unique_hashes = set(all_hashes)\n",
    "    assert len(unique_hashes) == len(genotypes), \"Genotypes must be, generally, unique\"\n",
    "\n",
    "    print(\"Unique genotypes: \", len(set(all_hashes)))\n",
    "    print(\"First genotype: \", genotypes[0])\n",
    "\n",
    "test_n_creation()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T16:11:15.976418Z",
     "end_time": "2023-05-11T16:12:45.867239Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes before mutation: \n",
      "{<expr3>: [0, 2, 0, 3, 3, 3, 3, 3, ...], <integer>: [0, 0, 2, 2, 1, 1, 0, 1, 0, 1, 2, 1, 2, 2, 2, 1, 2, 2, 1, 1, 0, 0, 1, 1, 0, 2, 2, 1, 0, 2, 1, 2, 1, 1, 2, 0, 2, 2, 1, 1, 0, 2, 0, 0, 1, 1, 0, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 0, 1, 0, 0, 2, 1, 2, 2, 2, 2, 1, 1, 2, 0, 0, 2, 0, 2, 1, 1, 0, 1, 0, ...], <non-zero-digit>: [7, 4, 0, 1, 4, 3, 6, 4, 2, 1, 2, 5, 0, 0, 6, 5, 1, 7, 7, 7, 3, 5, 4, 4, 5, 6, 0, 1, 0, 7, 6, 0, 7, 5, 3, 7, 0, 0, 3, 3, 4, 6, 1, 0, 8, 1, 3, 8, 5, 2, 1, 4, 3, 4, 7, 0, 6, 3, 8, 6, 7, 3, 7, 6, 6, 4, 2, 1, 3, 3, 3, 4, 5, 6, 3, 0, 2, 5, 1, 3, 5, 2, 7, 3, 4, 8, 4, 8, 0, 7, 5, 6, 2, 1, 6, 7, 8, 7, 7, 3, 6, 7, 6, 3, 0, 4, 0, 4, 2, 7, 4, 0, 3, 1, 7, 4, 8, 7, 4, 7, 4, 3, 5, 7, 2, 2, 3, 6, 0, 5, 7, 6, 3, 3, 8, 4, 1, 3, 6, 5, 0, 8, 7, 0, 2, 2, 3, 5, 0, 8, 5, 1, 7, 7, 1, 1, 5, 0, 2, 6, 7, 7, 0, 1, 4, 1, 4, 2, 1, 6, 6, 7, 0, 7, 6, 4, 3, 7, 8, 8, 1, 4, 6, 0, 0, 4, 0, 3, 7, 3, 2, 8, 0, 4, 4, 8, 6, 1, 2, 5, 5, 8, 2, 7, 6, 0, 4, 7, 0, 4, 3, 4, 1, 4, 3, 2, 3, 2, 1, 1, 3, 3, 3, 7, 6, 0, 5, 8, 1, 5, 8, 0, 8, 4, 5, 2, 8, 6, 1, 3, 2, 8, 2, 1, 1, 7, 3, 1, 8, 3, 2, 2, 5, 1, 0, 7, 1, 3, 2, 1, 7, 5, 0, 5, 2, 4, 6, 2, 4, 1, 6, 0, 3, 5, 0, 5, 7, 7, 5, 0, 2, 8, 3, 7, 8, 5, 1, 2, 5, 1, 0, 7, 5, 7, 2, 3, 2, 7, 3, 2, 5, 1, 6, 6, 4, 5, 0, 4, 4, 1, 7, 2, 5, 4, 1, 8, 1, 1, 7, 1, ...], <expr4>: [0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, ...], <var>: [0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, ...], <op>: [0, 2, 3, 1, 1, 1, 1, 0, 2, 1, 1, 1, 0, 3, 1, 1, 3, 0, 3, 2, 3, 3, 1, 3, 1, 1, 1, 3, 2, 2, 0, 0, ...], <digit>: [1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, ...], <expr>: [2, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...], <expr2>: [1, 3, 2, 1, ...], <expr1>: [1, 0, ...], <uop>: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...]}\n",
      "Hash:  g9vw29olnvhv\n",
      "\n",
      "\n",
      "Genotypes after mutation: \n",
      "{<expr3>: [4, 2, 0, 2, 3, 0, 3, 3, ...], <integer>: [0, 0, 2, 2, 1, 1, 0, 1, 0, 1, 2, 1, 2, 2, 2, 1, 2, 2, 1, 1, 0, 0, 1, 1, 0, 2, 2, 1, 0, 2, 1, 2, 1, 1, 2, 0, 2, 2, 1, 1, 0, 2, 0, 0, 1, 1, 0, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 0, 1, 0, 0, 2, 1, 2, 2, 2, 2, 1, 1, 2, 0, 0, 2, 0, 2, 1, 1, 0, 1, 0, ...], <non-zero-digit>: [7, 4, 0, 3, 4, 3, 9, 4, 6, 1, 0, 5, 0, 4, 6, 5, 1, 7, 3, 7, 4, 3, 0, 4, 7, 7, 0, 0, 2, 7, 6, 4, 7, 5, 4, 7, 0, 9, 2, 3, 4, 3, 1, 0, 8, 1, 3, 8, 2, 2, 1, 9, 3, 4, 1, 4, 6, 8, 1, 6, 4, 3, 3, 6, 0, 4, 2, 6, 1, 3, 0, 4, 1, 6, 6, 5, 7, 2, 1, 3, 4, 8, 7, 3, 8, 4, 0, 8, 5, 2, 0, 2, 2, 7, 6, 1, 8, 8, 1, 3, 6, 6, 6, 3, 5, 1, 0, 9, 2, 7, 9, 0, 3, 2, 7, 1, 1, 7, 4, 7, 4, 3, 8, 7, 7, 1, 8, 6, 0, 5, 7, 4, 4, 6, 8, 4, 1, 3, 6, 5, 0, 8, 7, 5, 1, 9, 3, 1, 9, 7, 5, 8, 6, 7, 2, 1, 5, 9, 2, 6, 7, 1, 0, 1, 4, 1, 4, 7, 1, 6, 0, 7, 0, 9, 6, 0, 3, 9, 8, 0, 3, 4, 6, 7, 0, 4, 0, 4, 0, 3, 6, 1, 4, 4, 1, 8, 4, 1, 2, 4, 5, 6, 9, 7, 6, 0, 4, 7, 9, 2, 3, 8, 9, 4, 3, 5, 7, 3, 7, 8, 6, 3, 8, 7, 5, 0, 5, 5, 6, 5, 0, 0, 8, 9, 5, 6, 0, 6, 1, 8, 1, 4, 1, 5, 1, 7, 4, 1, 8, 7, 2, 1, 4, 1, 1, 5, 0, 3, 2, 1, 7, 3, 0, 2, 2, 2, 7, 0, 4, 4, 8, 9, 3, 5, 0, 5, 6, 6, 5, 0, 2, 8, 7, 5, 8, 5, 1, 2, 6, 7, 0, 7, 6, 7, 5, 3, 9, 7, 2, 7, 5, 8, 6, 6, 2, 5, 2, 4, 7, 1, 6, 2, 5, 5, 1, 1, 7, 1, 7, 1, ...], <expr4>: [2, 0, 0, 0, 0, 0, 2, 1, 2, 2, 2, 0, 1, 2, 1, 0, ...], <var>: [0, 1, 2, 2, 0, 0, 2, 0, 1, 2, 0, 2, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 2, 0, 2, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 2, 1, 0, 1, 2, 2, 1, 2, 1, 2, 0, 1, 0, 0, 1, 1, 2, 0, 2, 2, 0, 0, 0, 2, 1, 2, 1, 2, 2, 2, 0, 1, 1, 1, 1, 2, 1, ...], <op>: [0, 0, 3, 0, 0, 2, 3, 4, 3, 2, 2, 0, 0, 1, 3, 1, 3, 1, 3, 2, 3, 3, 3, 3, 4, 3, 1, 3, 0, 2, 0, 4, ...], <digit>: [1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 2, 0, 1, 0, 2, 0, 0, 1, 2, 2, 1, 1, 0, 1, 1, 1, 1, 1, 1, 2, 1, 0, 0, 1, 1, 0, 2, 0, 1, 1, 1, 1, 1, 0, 0, 1, 2, 1, 0, 1, 1, 1, 2, 1, 0, 0, 0, 0, 2, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 2, 1, 2, 1, 1, 1, 0, 1, 1, 1, 2, 0, 2, 2, 2, 1, 1, 0, 0, 1, 1, 2, 2, 0, 2, 2, 2, 1, 1, 1, 2, 2, 0, 1, 0, 2, 2, 0, 0, 0, 1, 2, 0, 1, 1, 0, 2, 1, 2, 0, 0, 1, 0, 2, 2, 2, 2, 2, 0, 2, 1, 1, 2, 2, 1, 2, 2, 0, 0, 0, 1, 0, 1, 2, 0, 0, 2, 0, 2, 0, 2, 0, 2, ...], <expr>: [2, ...], <start>: [0, ...], <number>: [1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, ...], <expr2>: [1, 4, 2, 4, ...], <expr1>: [2, 4, ...], <uop>: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...]}\n",
      "Hash:  1cksciy6c1efx\n"
     ]
    }
   ],
   "source": [
    "GENOTYPE_TYPE = Dict.empty(\n",
    "    key_type=types.unicode_type,\n",
    "    value_type=production_type\n",
    ")\n",
    "\n",
    "def assert_equality_of_hashes(hash1, hash2):\n",
    "    if hash1 != hash2:\n",
    "        print(\"Hashes are not equal\")\n",
    "        print(\"Hash1: \", hash1)\n",
    "        print(\"Hash2: \", hash2)\n",
    "        raise ValueError(\"Hashes are not equal\")\n",
    "\n",
    "def mutate_genotypes_1(genotypes, grammar,  mutation_rate_operators: float):\n",
    "    for genotype_instance in genotypes:\n",
    "        mutate_genotype_inplace(genotype_instance, grammar, mutation_rate_operators)\n",
    "\n",
    "\n",
    "def mutate_genotypes_2(genotypes, grammar, mutation_rate_items: float, mutation_rate_operators: float):\n",
    "    for genotype in genotypes:\n",
    "        # Pick random genotypes to mutate, using the mutation_rate_items.\n",
    "        if rand() < mutation_rate_items:\n",
    "            mutate_genotype_inplace(genotype, grammar, mutation_rate_operators)\n",
    "\n",
    "def mutate_genotypes_3(genotypes, grammar, mutation_rate_items: float, mutation_rate_operators: float):\n",
    "    for genotype in genotypes:\n",
    "        # Pick random keys to mutate, using the mutation_rate_items\n",
    "        subgenotype_to_mutate = new_genotype_type.copy()\n",
    "        for key in genotype:\n",
    "            if random.random() < mutation_rate_items:\n",
    "                subgenotype_to_mutate[key] = genotype[key]\n",
    "            #[key for key in genotype if random.random() < mutation_rate_items]\n",
    "\n",
    "        mutate_genotype_inplace(subgenotype_to_mutate, grammar, mutation_rate_operators)\n",
    "\n",
    "        # Update the genotype with the mutated sub-genotype\n",
    "        for key in subgenotype_to_mutate:\n",
    "            genotype[key] = subgenotype_to_mutate[key]\n",
    "\n",
    "def mutate_genotype_old(mutable_genotype, grammar, mutation_rate_operators: float):\n",
    "    genotype_keys = list(mutable_genotype.keys())\n",
    "    for i in range(len(genotype_keys)):\n",
    "        print(\"Current genotype: \", mutable_genotype)\n",
    "        key = genotype_keys[i]\n",
    "        gene = mutable_genotype[key]\n",
    "        print(\"Current key: \", key)\n",
    "        print(\"Current gene: \", gene)\n",
    "\n",
    "        productions = grammar[key]\n",
    "        productions_length = len(productions)\n",
    "        for j in range(len(gene)):\n",
    "            # Decide whether to mutate this gene based on the mutation_rate_operators\n",
    "            if random.random() < mutation_rate_operators:\n",
    "                # Choose a random production index different from the current gene index\n",
    "                print(f\"Mutating gene {gene} at index {j}\")\n",
    "                new_production_index = random_int(0, productions_length - 1)\n",
    "                while new_production_index == gene[j]:\n",
    "                    new_production_index = random_int(0, productions_length - 1)\n",
    "                gene[j] = new_production_index\n",
    "\n",
    "        mutable_genotype[key] = gene\n",
    "\n",
    "\n",
    "@njit\n",
    "def mutate_genotype_inplace(mutable_genotype: Dict, grammar: Dict, mutation_rate_operators: float):\n",
    "    genotype_keys = List(mutable_genotype.keys())\n",
    "    count = 0\n",
    "    for i in range(len(genotype_keys)):\n",
    "        key = genotype_keys[i]\n",
    "        gene = mutable_genotype[key]\n",
    "\n",
    "        productions = grammar[key]\n",
    "        productions_length = len(productions)\n",
    "        for j in range(len(gene)):\n",
    "            # Decide whether to mutate this gene based on the mutation_rate_operators\n",
    "            if np.random.random() < mutation_rate_operators:\n",
    "                # Choose a random production index different from the current gene index\n",
    "                new_production_index = random_int(0, productions_length)\n",
    "                attempts = 0\n",
    "                max_attempts = 10\n",
    "                while new_production_index == gene[j] and attempts < max_attempts:\n",
    "                    new_production_index = random_int(0, productions_length)\n",
    "                    attempts += 1\n",
    "                if attempts < max_attempts:\n",
    "                    gene[j] = new_production_index\n",
    "                    count += 1\n",
    "                else:\n",
    "                    print(f\"At key {key}: Could not mutate gene {gene} at index {j} after {max_attempts} attempts, only have {productions_length} productions: {productions}\")\n",
    "\n",
    "        mutable_genotype[key] = gene\n",
    "    return count\n",
    "\n",
    "def test_mutation():\n",
    "    genotypes, grammar = create_n_genotypes(1, 5, 2)\n",
    "    print(\"Genotypes before mutation: \")\n",
    "    print(genotypes[0])\n",
    "    print(\"Hash: \", genotype_hash(genotypes[0]))\n",
    "\n",
    "    mutate_genotypes_3(genotypes, grammar, 0.5, 0.5)\n",
    "\n",
    "    print(\"\\n\\nGenotypes after mutation: \")\n",
    "    print(genotypes[0])\n",
    "    print(\"Hash: \", genotype_hash(genotypes[0]))\n",
    "\n",
    "test_mutation()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T19:32:26.179116Z",
     "end_time": "2023-05-11T19:32:47.598778Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes before crossover: \n",
      "{<integer>: [0, 2, 1, 1, ...], <non-zero-digit>: [0, 4, 5, 2, 3, 3, 5, 3, 4, 4, 3, 1, 0, 2, 6, 3, ...], <var>: [0, 1, 1, 1, ...], <op>: [3, ...], <digit>: [1, 1, 0, 1, 1, 1, 0, 0, ...], <expr>: [3, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, ...], <expr1>: [0, 1, ...], <uop>: [0, ...]}\n",
      "{<integer>: [0, 2, 2, 1, ...], <non-zero-digit>: [7, 3, 1, 7, 1, 4, 1, 4, 0, 0, 0, 6, 7, 7, 3, 8, ...], <var>: [1, 1, 1, 1, ...], <op>: [3, ...], <digit>: [0, 1, 0, 0, 1, 1, 1, 0, ...], <expr>: [1, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, ...], <expr1>: [1, 0, ...], <uop>: [0, ...]}\n",
      "_______\n",
      "Finished crossover\n",
      "Genotypes after crossover: \n",
      "{<integer>: [0, 2, 1, 1, ...], <non-zero-digit>: [7, 3, 1, 7, 1, 4, 1, 4, 0, 0, 0, 6, 7, 7, 3, 8, ...], <var>: [1, 1, 1, 1, ...], <op>: [3, ...], <digit>: [1, 1, 0, 1, 1, 1, 0, 0, ...], <expr>: [1, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, ...], <expr1>: [0, 1, ...], <uop>: [0, ...]}\n",
      "{<integer>: [0, 2, 2, 1, ...], <non-zero-digit>: [0, 4, 5, 2, 3, 3, 5, 3, 4, 4, 3, 1, 0, 2, 6, 3, ...], <var>: [0, 1, 1, 1, ...], <op>: [3, ...], <digit>: [0, 1, 0, 0, 1, 1, 1, 0, ...], <expr>: [3, ...], <start>: [0, ...], <number>: [0, 0, 0, 0, ...], <expr1>: [1, 0, ...], <uop>: [0, ...]}\n",
      "_______\n"
     ]
    }
   ],
   "source": [
    "@njit\n",
    "def create_mask(length: int, crossover_probability: float) -> List[int]:\n",
    "    mask = List.empty_list(types.int64)\n",
    "    for _ in range(length):\n",
    "        mask.append(1 if random.random() < crossover_probability else 0)\n",
    "    #print(mask)\n",
    "    return mask\n",
    "\n",
    "\n",
    "@njit\n",
    "def crossover_numba(a: Dict, b: Dict, crossover_probability: float, genotype_type=new_genotype_type) -> List[Dict]:\n",
    "    mask = create_mask(len(a), crossover_probability)\n",
    "\n",
    "    child1 = a.copy()\n",
    "    child2 = a.copy()\n",
    "\n",
    "    idx = 0\n",
    "    for key in a.keys():\n",
    "        if mask[idx] == 1:\n",
    "            child1[key] = a[key]\n",
    "            child2[key] = b[key]\n",
    "        else:\n",
    "            child1[key] = b[key]\n",
    "            child2[key] = a[key]\n",
    "        idx += 1\n",
    "\n",
    "    result = list([child1, child2])\n",
    "    print(\"Finished crossover\")\n",
    "    return result\n",
    "\n",
    "def print_genotypes(genotypes):\n",
    "    for genotype in genotypes:\n",
    "        print(genotype)\n",
    "    print(\"_______\")\n",
    "\n",
    "\n",
    "def test_crossover():\n",
    "    n = 2\n",
    "    depth = 2\n",
    "\n",
    "    genotypes_test, grammar = create_n_genotypes(n, depth, 2)\n",
    "    print(\"Genotypes before crossover: \")\n",
    "    print_genotypes(genotypes_test[0:2])\n",
    "    hash_a_before = genotype_hash(genotypes_test[0])\n",
    "    hash_b_before = genotype_hash(genotypes_test[1])\n",
    "\n",
    "    new_genotypes = crossover_numba(genotypes_test[0], genotypes_test[1], 0.5, new_genotype_type)\n",
    "\n",
    "    print(\"Genotypes after crossover: \")\n",
    "    print_genotypes(new_genotypes)\n",
    "\n",
    "    hash_a_after = genotype_hash(genotypes_test[0])\n",
    "    hash_b_after = genotype_hash(genotypes_test[1])\n",
    "\n",
    "\n",
    "    assert_equality_of_hashes(hash_a_before, hash_a_after)\n",
    "    assert_equality_of_hashes(hash_b_before, hash_b_after)\n",
    "\n",
    "test_crossover()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T19:33:56.917710Z",
     "end_time": "2023-05-11T19:34:10.091217Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "outputs": [],
   "source": [
    "@njit\n",
    "def create_full_tree_from_genome(genotype, grammar, node_value_type,  print_tree=False) -> Dict:\n",
    "    new_nodes = Dict.empty(key_type=types.int64, value_type=node_value_type)\n",
    "    create_node('<start>', new_nodes)\n",
    "    load_genotype(new_nodes, genotype, grammar)\n",
    "    if print_tree:\n",
    "        print(simple_repr(0, new_nodes))\n",
    "    return new_nodes\n",
    "\n",
    "@njit\n",
    "def create_and_eval_tree(genotype, grammar, variables, node_value_type):\n",
    "    new_nodes = create_full_tree_from_genome(genotype, grammar, node_value_type, print_tree=True)\n",
    "    print(evaluate(0, new_nodes, variables))\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:12:49.227191Z",
     "end_time": "2023-05-11T15:12:49.227191Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{x: 42.0, y: 10.0}\n",
      "y\n",
      "10.0\n",
      "_______\n",
      "x\n",
      "42.0\n",
      "_______\n",
      "( '/' ( 'abs' ( 1 1 ) ) ( 'abs' y ) )\n",
      "1.1\n",
      "_______\n",
      "( 'abs' ( '+' y ( '+' ( 'abs' x ) ( 6 0 ) ) ) )\n",
      "112.0\n",
      "_______\n",
      "( '*' y ( 8 ) )\n",
      "80.0\n",
      "_______\n",
      "( '+' ( 'abs' y ) ( 7 ) )\n",
      "17.0\n",
      "_______\n",
      "y\n",
      "10.0\n",
      "_______\n",
      "( '/' ( 'abs' ( '*' ( 'abs' x ) ( 0 ) ) ) ( 9 3 ) )\n",
      "0.0\n",
      "_______\n",
      "( 'abs' x )\n",
      "42.0\n",
      "_______\n",
      "( 'abs' ( '*' x ( 'abs' ( '+' ( 0 ) x ) ) ) )\n",
      "1764.0\n",
      "_______\n",
      "x\n",
      "42.0\n",
      "_______\n",
      "x\n",
      "42.0\n",
      "_______\n",
      "y\n",
      "10.0\n",
      "_______\n",
      "( '+' ( 'abs' ( '+' ( 6 0 9 ) y ) ) ( '/' ( 1 0 0 ) ( 1 ) ) )\n",
      "719.0\n",
      "_______\n",
      "( '*' ( 'abs' ( 'abs' ( '*' ( 0 ) ( 0 ) ) ) ) ( 'abs' ( 0 ) ) )\n",
      "0.0\n",
      "_______\n",
      "y\n",
      "10.0\n",
      "_______\n",
      "( '+' ( '+' ( '+' ( 'abs' ( 8 0 0 ) ) ( 4 7 4 ) ) ( 3 0 0 ) ) ( 'abs' ( '*' ( 'abs' y ) ( 'abs' ( 8 6 9 ) ) ) ) )\n",
      "10264.0\n",
      "_______\n",
      "x\n",
      "42.0\n",
      "_______\n",
      "x\n",
      "42.0\n",
      "_______\n",
      "( 2 0 3 )\n",
      "203.0\n",
      "_______\n"
     ]
    }
   ],
   "source": [
    "@njit\n",
    "def create_variable_dict(variable_names, variable_values):\n",
    "    variables = Dict.empty(key_type=types.unicode_type, value_type=types.float64)\n",
    "    for i in range(len(variable_names)):\n",
    "        variables[variable_names[i]] = variable_values[i]\n",
    "    return variables\n",
    "\n",
    "\n",
    "NODE_TYPE = NodeData.class_type.instance_type\n",
    "\n",
    "variable_2 = create_variable_dict(['x', 'y'], [42, 10])\n",
    "print(variable_2)\n",
    "\n",
    "\n",
    "def test_evaluate(variables, max_depth=5 ):\n",
    "    iterations = 20\n",
    "    variable_count = len(variables)\n",
    "    genotypes, grammar = create_n_genotypes(iterations, max_depth, variable_count)\n",
    "\n",
    "    for genotype in genotypes:\n",
    "        create_and_eval_tree(genotype, grammar, variables, NODE_TYPE)\n",
    "        print(\"_______\")\n",
    "\n",
    "test_evaluate(variable_2, 5)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T19:35:41.996223Z",
     "end_time": "2023-05-11T19:36:40.819803Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitness: 1hx0gqkra2l9i and copy 1hx0gqkra2l9i and empty 0\n",
      "After fitness: 1hx0gqkra2l9i and copy ym3vdfhcaqzg and empty 0\n",
      "y_predicted: 4.0, y_correct: 1\n",
      "y_predicted: 6.0, y_correct: 2\n",
      "RMSE: 3.5355339059327378\n",
      "Before fitness: 1hx0gqkra2l9i and copy 1hx0gqkra2l9i and empty 0\n",
      "After fitness: 1hx0gqkra2l9i and copy ym3vdfhcaqzg and empty 0\n",
      "y_predicted: 4.0, y_correct: 1\n",
      "y_predicted: 6.0, y_correct: 2\n",
      "RMSE: 3.5355339059327378\n"
     ]
    }
   ],
   "source": [
    "@njit\n",
    "def create_variable_names(variable_count):\n",
    "    variable_names = List.empty_list(types.unicode_type)\n",
    "    # start at x, y,z...\n",
    "    for i in range(variable_count):\n",
    "        variable_names.append(chr(ord('x') + i))\n",
    "    return variable_names\n",
    "\n",
    "@njit\n",
    "def calculate_fitness(variables_values, y_values, genotype, grammar, node_value_type, EMPTY_GENOTYPE, print_tree=False):\n",
    "    empty_genotype = EMPTY_GENOTYPE.copy()\n",
    "\n",
    "    empty_genotype_hash = genotype_hash(EMPTY_GENOTYPE)\n",
    "    new_genotype = deep_copy_genotype(genotype, empty_genotype)\n",
    "\n",
    "    param_hash = genotype_hash(genotype)\n",
    "    new_genotype_hash = genotype_hash(new_genotype)\n",
    "    report = \"Before fitness: \" + str(param_hash) + \" and copy \" + str(new_genotype_hash) + \" and empty \" + str(empty_genotype_hash)\n",
    "    print(report)\n",
    "    new_nodes = create_full_tree_from_genome(new_genotype, grammar, node_value_type)\n",
    "\n",
    "    param_hash = genotype_hash(genotype)\n",
    "    new_genotype_hash = genotype_hash(new_genotype)\n",
    "    empty_genotype_hash = genotype_hash(EMPTY_GENOTYPE)\n",
    "\n",
    "    report = \"After fitness: \" + str(param_hash) + \" and copy \" + str(new_genotype_hash) + \" and empty \" + str(empty_genotype_hash)\n",
    "    print(report)\n",
    "    if print_tree:\n",
    "        print(simple_repr(0, new_nodes))\n",
    "        print(\"_______\")\n",
    "    # create variables\n",
    "    variable_count = len(variables_values)\n",
    "    variable_names = create_variable_names(variable_count)\n",
    "\n",
    "    evaluation_cases = len(y_values)\n",
    "\n",
    "    sum = 0\n",
    "    results = List.empty_list(types.float64)\n",
    "    for i in range(evaluation_cases):\n",
    "        current_variable_values = []\n",
    "        for j in range(variable_count):\n",
    "            current_variable_values.append(variables_values[j][i])\n",
    "\n",
    "        variable_dict = create_variable_dict(variable_names, current_variable_values)\n",
    "        y_predicted = evaluate(0, new_nodes, variable_dict)\n",
    "\n",
    "        y_correct = y_values[i]\n",
    "        squared_diff = (y_predicted - y_correct) ** 2\n",
    "        results.append(y_predicted)\n",
    "        sum += squared_diff\n",
    "\n",
    "    rmse = np.sqrt(sum / evaluation_cases)\n",
    "    return rmse, results\n",
    "\n",
    "\n",
    "genotype, grammar = create_genotype(max_depth=5, variables_count=2)\n",
    "\n",
    "test_data = np.array([\n",
    "    [1, 2],\n",
    "    [3, 4]\n",
    "], dtype=np.float64)\n",
    "test_y = [1, 2]\n",
    "\n",
    "\n",
    "@njit\n",
    "def create_empty_dict_with_same_type(a):\n",
    "    key_type = a.key_type\n",
    "    value_type = a.value_type\n",
    "\n",
    "    empty_dict = Dict.empty(key_type=key_type, value_type=value_type)\n",
    "    return empty_dict\n",
    "\n",
    "genotype_test, grammar = create_genotype(max_depth=5, variables_count=2)\n",
    "\n",
    "def test_fitness(genotype_test):\n",
    "    # Call the calculate_fitness function\n",
    "    rmse, y_predicted_values = calculate_fitness(test_data, test_y, genotype_test, grammar, NODE_TYPE, GENOTYPE_TYPE)\n",
    "\n",
    "    # Print the values\n",
    "    for y_pred, y_corr in zip(y_predicted_values, test_y):\n",
    "        print(f\"y_predicted: {y_pred}, y_correct: {y_corr}\")\n",
    "\n",
    "    print(f\"RMSE: {rmse}\")\n",
    "\n",
    "test_fitness(genotype_test)\n",
    "print(\"Should return equal values and hashes...\")\n",
    "test_fitness(genotype_test)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-11T15:54:16.522075Z",
     "end_time": "2023-05-11T15:54:34.141240Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "@njit(parallel=True)\n",
    "def calculate_all_fitnesses(genotypes: List[dict],\n",
    "                            variables_values: np.ndarray,\n",
    "                            y_values: np.ndarray,\n",
    "                            grammar: dict,\n",
    "                            node_value_type: types.Type) -> np.ndarray:\n",
    "\n",
    "    num_genotypes = len(genotypes)\n",
    "    fitness_list = np.empty(num_genotypes, dtype=np.float64)\n",
    "\n",
    "    for i in prange(num_genotypes):\n",
    "        genotype = genotypes[i].copy()\n",
    "        rmse, _ = calculate_fitness(variables_values, y_values, genotype, grammar, node_value_type)\n",
    "        fitness_list[i] = rmse\n",
    "\n",
    "    return fitness_list\n",
    "\n",
    "# Example usage\n",
    "genotypes, grammar = create_n_genotypes(20, 5, 2)\n",
    "test_data = np.array([\n",
    "    [1, 2],\n",
    "    [3, 4]\n",
    "], dtype=np.float64)\n",
    "test_y = [1, 2]\n",
    "\n",
    "fitness_list = calculate_all_fitnesses(genotypes, test_data, test_y, grammar, NODE_TYPE)\n",
    "\n",
    "print(\"Fitness list: \", fitness_list)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-10T19:34:26.177029Z",
     "end_time": "2023-05-10T19:34:57.615852Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from numba import njit, prange\n",
    "from numba import types\n",
    "\n",
    "@njit(parallel=True)\n",
    "def selection_tournament(fitness_list, tournament_size, maximize=False, print_groups=False):\n",
    "    shuffled_indices = np.arange(len(fitness_list))\n",
    "    np.random.shuffle(shuffled_indices)\n",
    "\n",
    "    num_groups = len(fitness_list) // tournament_size\n",
    "    extra_group = len(fitness_list) % tournament_size != 0\n",
    "\n",
    "    best_elements = np.empty(num_groups, dtype=np.int64)\n",
    "    group_strings = np.empty(num_groups, dtype=types.unicode_type)\n",
    "\n",
    "    for group_idx in prange(num_groups):\n",
    "        start_idx = group_idx * tournament_size\n",
    "        end_idx = start_idx + tournament_size\n",
    "        # account for the last partial group being smaller and unaccounted, so we merge it with the last full group\n",
    "        if group_idx == num_groups - 1 and extra_group:\n",
    "            end_idx = len(fitness_list)\n",
    "        group_string = \"[\"\n",
    "        best_in_group = shuffled_indices[start_idx]\n",
    "        group_string += str(fitness_list[best_in_group]) + \" \"\n",
    "        for idx in range(start_idx + 1, end_idx):\n",
    "            correct_index = shuffled_indices[idx]\n",
    "            group_string += str(fitness_list[correct_index]) + \" \"\n",
    "            if maximize:\n",
    "                if fitness_list[correct_index] > fitness_list[best_in_group]:\n",
    "                    best_in_group = shuffled_indices[idx]\n",
    "            else:\n",
    "                if fitness_list[correct_index] < fitness_list[best_in_group]:\n",
    "                    best_in_group = shuffled_indices[idx]\n",
    "        group_string += \"]\"\n",
    "        group_strings[group_idx] = group_string\n",
    "        best_elements[group_idx] = best_in_group\n",
    "\n",
    "\n",
    "    if print_groups:\n",
    "        for group_string in group_strings:\n",
    "            print(group_string)\n",
    "    return best_elements\n",
    "\n",
    "@njit(parallel=True)\n",
    "def best_n_items(fitness_list, n, maximize=False):\n",
    "    if maximize:\n",
    "        sorted_indices = np.argsort(-fitness_list)\n",
    "    else:\n",
    "        sorted_indices = np.argsort(fitness_list)\n",
    "    return sorted_indices[:n]\n",
    "\n",
    "# Example usage\n",
    "fitness_list = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 99, 98, 97, 96, 95, 94, 93, 92, 91])\n",
    "tournament_size = 6\n",
    "best_elements = selection_tournament(fitness_list, tournament_size, maximize=False, print_groups=True)\n",
    "\n",
    "# Get the value of best elements\n",
    "best_elements_selection = [fitness_list[idx] for idx in best_elements]\n",
    "print(\"Best elements: \", best_elements_selection)\n",
    "best_itens = best_n_items(fitness_list, 3, maximize=False)\n",
    "\n",
    "best_elements_global = [fitness_list[idx] for idx in best_itens]\n",
    "print(\"Global best elements: \", best_elements_global )"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-10T19:13:02.088538Z",
     "end_time": "2023-05-10T19:13:12.521356Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def print_resumed_genotype(genotype: Dict):\n",
    "\n",
    "    # calculate an unique hash for the genotype\n",
    "    hash_genotype = genotype_hash(genotype)\n",
    "\n",
    "    print(f\"{hash_genotype}: ______\")\n",
    "    keys = list(genotype.keys())\n",
    "    alphabetical_keys = sorted(keys)\n",
    "    for key in alphabetical_keys:\n",
    "        value = genotype[key]\n",
    "        real_len = min(len(value), 4)\n",
    "        print(f\"{key}: {value[:real_len]}\", end=\" \")\n",
    "\n",
    "def mutate_genotype(parent_genotype, grammar, probability ):\n",
    "    parameter_hash = genotype_hash(parent_genotype)\n",
    "\n",
    "    offspring_genotype = parent_genotype.copy()\n",
    "\n",
    "    offspring_genotype_hash = genotype_hash(offspring_genotype)\n",
    "\n",
    "    print(f\"pre mutation: Parent hash: {parameter_hash} Offspring hash: {offspring_genotype_hash}\")\n",
    "\n",
    "    #print(\"Offspring genotype: \", offspring_genotype)\n",
    "    mutate_genotype_inplace(offspring_genotype, grammar, probability)\n",
    "\n",
    "    offspring_genotype_hash = genotype_hash(offspring_genotype)\n",
    "    parameter_hash = genotype_hash(parent_genotype)\n",
    "\n",
    "    print(f\"post mutation: Parent hash: {parameter_hash} Offspring hash: {offspring_genotype_hash}\")\n",
    "    #print(\"Parent genotype: \", parent_genotype)\n",
    "   # raise Exception(\"Stop\")\n",
    "    return offspring_genotype\n",
    "\n",
    "def next_generation(genotype_list,\n",
    "                    grammar: dict,\n",
    "                    p_mutation: float,\n",
    "                    p_crossover: float,\n",
    "                    elite_percentage: float,\n",
    "                    variables_values: np.ndarray,\n",
    "                    y_values: np.ndarray,\n",
    "                    tournament_size: int):\n",
    "\n",
    "    population_size = len(genotype_list)\n",
    "\n",
    "    def calculate_single_fitness(genotype):\n",
    "        return calculate_fitness(variables_values, y_values, genotype, grammar, NODE_TYPE)\n",
    "\n",
    "    # Calculate fitness for all genotypes\n",
    "    fitness_list = calculate_all_fitnesses(genotype_list, variables_values, y_values, grammar, NODE_TYPE)\n",
    "    print(\"After fitness genotypes: \")\n",
    "\n",
    "    for i in range(len(genotype_list)):\n",
    "        print(\"\\nGenotype: \", i)\n",
    "        print_resumed_genotype(genotype_list[i])\n",
    "    print(\"\\n\\n\\n\")\n",
    "\n",
    "    unique, counts = np.unique(fitness_list, return_counts=True)\n",
    "\n",
    "\n",
    "\n",
    "    # Count number of repeated items\n",
    "    num_repeated = np.sum(counts[counts > 1]) - len(counts[counts > 1]) + len(unique[counts == 1])\n",
    "\n",
    "    # Metrics: Best and worst fitness\n",
    "    best_fitness = np.min(fitness_list)\n",
    "    worst_fitness = np.max(fitness_list)\n",
    "\n",
    "    # Metrics: Average fitness\n",
    "    avg_fitness = np.mean(fitness_list)\n",
    "\n",
    "\n",
    "    # Select the elite individuals\n",
    "    num_elite = int(population_size * elite_percentage)\n",
    "    elite_indices = best_n_items(fitness_list, num_elite)\n",
    "    elite_genotypes = [genotype_list[idx] for idx in elite_indices]\n",
    "\n",
    "    # Tournament selection for survivors\n",
    "    survivors_indices = selection_tournament(fitness_list, tournament_size)\n",
    "\n",
    "    survivor_hashes = [genotype_hash(genotype_list[idx]) for idx in survivors_indices]\n",
    "\n",
    "    survivors_genotypes = [genotype_list[idx] for idx in survivors_indices]\n",
    "\n",
    "\n",
    "    worse_than_father = 0\n",
    "    better_than_father = 0\n",
    "\n",
    "    print(\"Survivors: \", len(survivors_genotypes))\n",
    "    # Reproduce, mutate, and crossover to create new offspring\n",
    "    new_genotypes = []\n",
    "    while len(new_genotypes) < population_size - num_elite:\n",
    "        r = np.random.random()\n",
    "        parent_idx = np.random.randint(0, len(survivors_genotypes))\n",
    "        parent_genotype = survivors_genotypes[parent_idx]\n",
    "        print_resumed_genotype(parent_genotype)\n",
    "\n",
    "        print(f\"Hash for parent: {survivor_hashes[parent_idx]}\")\n",
    "\n",
    "        if r < p_mutation:\n",
    "            print(\"Mutating...\")\n",
    "            # Mutate\n",
    "            offspring_genotype = mutate_genotype(parent_genotype, grammar, p_mutation)\n",
    "        elif r < p_mutation + p_crossover:\n",
    "            print(\"Crossover...\")\n",
    "            # Crossover\n",
    "            parent2_idx = np.random.randint(0, len(survivors_genotypes))\n",
    "            attemps = 0\n",
    "            max_attemps = 10\n",
    "            while parent2_idx == parent_idx and attemps < max_attemps:\n",
    "                parent2_idx = np.random.randint(0, len(survivors_genotypes))\n",
    "                attemps += 1\n",
    "            if attemps == max_attemps:\n",
    "                print(\"Couldn't find a different parent to crossover, just mutating\")\n",
    "                # If we couldn't find a different parent, just mutate\n",
    "                offspring_genotype = mutate_genotype(parent_genotype, grammar, p_mutation)\n",
    "            else:\n",
    "                print(f\"Hash for parent: {survivor_hashes[parent_idx]}\")\n",
    "\n",
    "                parent2_genotype = survivors_genotypes[parent2_idx]\n",
    "                print_resumed_genotype(parent_genotype)\n",
    "                print_resumed_genotype(parent2_genotype)\n",
    "                print(\"++++++++++ end parents\")\n",
    "                offspring_1, offspring_2 = crossover_numba(parent_genotype, parent2_genotype, p_crossover, GENOTYPE_TYPE)\n",
    "\n",
    "                print_resumed_genotype(offspring_1)\n",
    "                print_resumed_genotype(offspring_2)\n",
    "\n",
    "                parents_mean_fitness = (fitness_list[parent_idx] + fitness_list[parent2_idx]) / 2\n",
    "\n",
    "                offspring_1_fitness = calculate_single_fitness(offspring_1)\n",
    "                offspring_2_fitness = calculate_single_fitness(offspring_2)\n",
    "                print(\"Created offspring with fitness: \", offspring_1_fitness, \" and \", offspring_2_fitness)\n",
    "                if offspring_1_fitness < parents_mean_fitness:\n",
    "                    better_than_father += 1\n",
    "                else:\n",
    "                    worse_than_father += 1\n",
    "                if offspring_2_fitness < parents_mean_fitness:\n",
    "                    better_than_father += 1\n",
    "                else:\n",
    "                    worse_than_father += 1\n",
    "\n",
    "                offspring_genotype = offspring_1\n",
    "                new_genotypes.append(offspring_2)\n",
    "        else:\n",
    "            print(\"Reproducing...\")\n",
    "            # Reproduce\n",
    "            offspring_genotype = parent_genotype.copy()\n",
    "\n",
    "        new_genotypes.append(offspring_genotype)\n",
    "\n",
    "    # Combine elite and new offspring to create the next generation\n",
    "    next_gen = elite_genotypes + new_genotypes\n",
    "    print(\"Better than father: \", better_than_father)\n",
    "    print(\"Worse than father: \", worse_than_father)\n",
    "    print(\"Num repeated: \", num_repeated)\n",
    "    print(\"Best fitness: \", best_fitness)\n",
    "    print(\"Worst fitness: \", worst_fitness)\n",
    "    print(\"Avg fitness: \", avg_fitness)\n",
    "    print(\"Num elite: \", num_elite)\n",
    "    print(\"Next gen length: \", len(next_gen))\n",
    "    return next_gen\n",
    "\n",
    "\n",
    "genotypes, grammar = create_n_genotypes(4, 3, 2)\n",
    "\n",
    "print(\"Original genotypes: \")\n",
    "for i in range(len(genotypes)):\n",
    "    print(\"\\nGenotype: \", i)\n",
    "    print_resumed_genotype(genotypes[i])\n",
    "print(\"\\n\\n\\n\")\n",
    "test_data = np.array([\n",
    "    [1, 2, 3],\n",
    "    [3, 4, 5]\n",
    "], dtype=np.float64)\n",
    "test_y = [10, 20, 30]\n",
    "\n",
    "next_gen = next_generation(genotypes, grammar, 0, 1, 0, test_data, test_y, 2)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "start_time": "2023-05-10T19:29:23.887972Z",
     "end_time": "2023-05-10T19:29:29.507631Z"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  },
  "colab": {
   "provenance": []
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
